File: orchalanguage/src/main/java/orcha/lang/compiler/referenceimpl/springIntegration/WhenInstructionForSpringIntegration.java
package orcha.lang.compiler.referenceimpl.springIntegration;

import orcha.lang.compiler.OrchaCompilationException;
import orcha.lang.compiler.WhenInstruction;

public class WhenInstructionForSpringIntegration extends WhenInstruction {

	public WhenInstructionForSpringIntegration() {
		super();
	}

	public WhenInstructionForSpringIntegration(String whenInstruction ) {
		super(whenInstruction);
	}
	
	public WhenInstructionForSpringIntegration(String whenInstruction, int instructionID, int lineNumber) {
		super(whenInstruction, instructionID, lineNumber);
	}
	
	public String convert(ApplicationOrEventInExpression application) {
		if(application.getState() == State.TERMINATES) {
			return "([" + (application.getOrder()-1) + "].payload instanceof Transpiler(orcha.lang.configuration.Application) AND [" + (application.getOrder()-1) + "].payload.state==Transpiler(orcha.lang.configuration.State).TERMINATED)";
		} else if(application.getState() == State.RECEIVES){
			return "([" + (application.getOrder()-1) + "].payload instanceof Transpiler(orcha.lang.configuration.EventHandler))";
		} else {
			return "([" + (application.getOrder()-1) + "].payload instanceof Transpiler(orcha.lang.configuration.Application) AND [" + (application.getOrder()-1) + "].payload.state==Transpiler(orcha.lang.configuration.State).FAILED)";
		}
	}
	
	public String logicalOperator(String operator) throws OrchaCompilationException {		
		if(operator.toUpperCase().equals("OR")) {
			return "OR";
		} else if(operator.toUpperCase().equals("AND")) {
			return "AND";
		} else {
			throw new OrchaCompilationException("Forbiden operator: " + operator + ". Only AND or OR are allowed.");
		}
	}

}

File: orchalanguage-spring-integration-implementation/src/main/kotlin/orcha/lang/compiler/referenceimpl/springIntegration/WhenInstructionForSpringIntegration.kt
package orcha.lang.compiler.referenceimpl.springIntegration

import orcha.lang.compiler.OrchaCompilationException
import orcha.lang.compiler.syntax.WhenInstruction

class WhenInstructionForSpringIntegration : WhenInstruction {

    constructor() : super() {}

    constructor(whenInstruction: String) : super(whenInstruction) {}

    constructor(whenInstruction: String, lineNumber: Int) : super(whenInstruction, lineNumber) {}

    override fun convert(application: WhenInstruction.ApplicationOrEventInExpression): String {
        return if (application.state == WhenInstruction.State.TERMINATES) {
            "([" + (application.order - 1) + "].payload instanceof Transpiler(orcha.lang.App) AND [" + (application.order - 1) + "].payload.state==Transpiler(orcha.lang.configuration.State).TERMINATED)"
        } else if (application.state == WhenInstruction.State.RECEIVES) {
            "([" + (application.order - 1) + "].payload instanceof Transpiler(orcha.lang.EventHandler))"
        } else {
            "([" + (application.order - 1) + "].payload instanceof Transpiler(orcha.lang.App) AND [" + (application.order - 1) + "].payload.state==Transpiler(orcha.lang.configuration.State).FAILED)"
        }
    }

    @Throws(OrchaCompilationException::class)
    override fun logicalOperator(operator: String): String {
        return if (operator.toUpperCase() == "OR") {
            "OR"
        } else if (operator.toUpperCase() == "AND") {
            "AND"
        } else {
            throw OrchaCompilationException("Forbiden operator: $operator. Only AND or OR are allowed.")
        }
    }

}
